app/api/internal/applications.rb

```diff
@@ -18,17 +18,25 @@ module API
           end
           app_attrs[:service_id] = service_id
           app_attrs[:id] = id
-          app = Application.save(app_attrs)
-          [201, headers, { status: :created, application: app.to_hash }.to_json]
+          begin
+            app = Application.save(app_attrs)
+            [201, headers, { status: :created, application: app.to_hash }.to_json]
+          rescue ApplicationHasNoState
+            [400, headers, { status: :bad_request, error: 'application cannot be created, it has no state' }.to_json]
+          end
         end
 
         put '/:id' do |service_id, id|
           app_attrs = api_params Application
           modified = Application.exists?(service_id, id)
           app_attrs[:service_id] = service_id
           app_attrs[:id] = id
-          app = Application.save(app_attrs)
-          { status: modified ? :modified : :created, application: app.to_hash }.to_json
+          begin
+            app = Application.save(app_attrs)
+            { status: modified ? :modified : :created, application: app.to_hash }.to_json
+          rescue ApplicationHasNoState
```

Comment:
Why should not be catched by exception_catcher middleware?

Reply:
because this is the internal API - it does not run the exception catcher and if it did that would be broken because the exception catcher responds with XML stuff while this API should respond JSON

Reply:
Then, it would be nice to have its own exception catcher processing responds as json bodies. Makes application logic easier to develop and more readable.